app/scripts/controllers/transactions/index.js

```diff
@@ -331,6 +331,137 @@ export default class TransactionController extends EventEmitter {
     });
```

Comment:
General comment:

When updating a transaction, this is the general pattern that will have to be followed:

```
const txMeta = this._getTransaction(txId);
txMeta.propertyA = newValueForA;
txMeta.propertyB = newValueForB;
this.txStateManager.updateTransaction(txMeta, note);
```

Reply:
I can't see the following properties here

gasLimit
gasPrice
estimateUsed
decEstimateBaseFee
sourceTokenSymbol
destinationTokenSymbol
destinationTokenDecimals
destinationTokenAddress
swapMetaData
swapTokenValue
userEditedGasLimit

I think the question is are these properties of txMeta or txMeta.txParams

@danjm 

> Also, for reference, here is an example txMeta object:
> 
> ```
> {
>   "id": 655970127251907,
>   "time": 1642690106761,
>   "status": "approved",
>   "metamaskNetworkId": "4",
>   "chainId": "0x4",
>   "loadingDefaults": false,
>   "dappSuggestedGasFees": null,
>   "txParams": {
>     "from": "0xb16f2b68800ea8adae005216ff3604e38e3884a6",
>     "to": "0xe18035bf8712672935fdb4e5e431b1a0183d2dfc",
>     "nonce": "0x28",
>     "value": "0x0",
>     "gas": "0x5208",
>     "maxFeePerGas": "0x59682f2a",
>     "maxPriorityFeePerGas": "0x59682f00",
>     "type": "0x2"
>   },
>   "origin": "metamask",
>   "type": "simpleSend",
>   "history": [
>     {
>       "id": 655970127251907,
>       "time": 1642690106761,
>       "status": "unapproved",
>       "metamaskNetworkId": "4",
>       "chainId": "0x4",
>       "loadingDefaults": true,
>       "dappSuggestedGasFees": null,
>       "txParams": {
>         "from": "0xb16f2b68800ea8adae005216ff3604e38e3884a6",
>         "to": "0xe18035bf8712672935fdb4e5e431b1a0183d2dfc",
>         "value": "0x0",
>         "gas": "0x5208",
>         "maxFeePerGas": "0x59682f2b",
>         "maxPriorityFeePerGas": "0x59682f00",
>         "type": "0x2"
>       },
>       "origin": "metamask",
>       "type": "simpleSend"
>     },
>     [
>       {
>         "op": "replace",
>         "path": "/loadingDefaults",
>         "value": false,
>         "note": "Added new unapproved transaction.",
>         "timestamp": 1642690106771
>       },
>       {
>         "op": "add",
>         "path": "/userFeeLevel",
>         "value": "medium"
>       }
>     ],
>     [
>       {
>         "op": "replace",
>         "path": "/txParams/maxFeePerGas",
>         "value": "0x59682f2a",
>         "note": "confTx: user approved transaction",
>         "timestamp": 1642690121224
>       },
>       {
>         "op": "add",
>         "path": "/estimatedBaseFee",
>         "value": "21"
>       }
>     ],
>     [
>       {
>         "op": "replace",
>         "path": "/status",
>         "value": "approved",
>         "note": "txStateManager: setting status to approved",
>         "timestamp": 1642690121224
>       }
>     ],
>     [
>       {
>         "op": "add",
>         "path": "/txParams/nonce",
>         "value": "0x28",
>         "note": "transactions#approveTransaction",
>         "timestamp": 1642690121236
>       },
>       {
>         "op": "add",
>         "path": "/nonceDetails",
>         "value": {
>           "params": {
>             "highestLocallyConfirmed": 0,
>             "highestSuggested": 40,
>             "nextNetworkNonce": 40
>           },
>           "local": {
>             "name": "local",
>             "nonce": 40,
>             "details": {
>               "startPoint": 40,
>               "highest": 40
>             }
>           },
>           "network": {
>             "name": "network",
>             "nonce": 40,
>             "details": {
>               "blockNumber": "0x98f680",
>               "baseCount": 40
>             }
>           }
>         }
>       }
>     ]
>   ],
>   "userFeeLevel": "medium",
>   "estimatedBaseFee": "21",
>   "nonceDetails": {
>     "params": {
>       "highestLocallyConfirmed": 0,
>       "highestSuggested": 40,
>       "nextNetworkNonce": 40
>     },
>     "local": {
>       "name": "local",
>       "nonce": 40,
>       "details": {
>         "startPoint": 40,
>         "highest": 40
>       }
>     },
>     "network": {
>       "name": "network",
>       "nonce": 40,
>       "details": {
>         "blockNumber": "0x98f680",
>         "baseCount": 40
>       }
>     }
>   }
> }
> ```

Reply:
Hi @danjm 

Maybe I don't need to worry much which property belongs to where. 

I can do this for instance in _updateTransaction

```
    const txMeta = this._getTransaction(txId);
    const updatedMeta = {...txMeta, ...txParams}
    this.txStateManager.updateTransaction(updatedMeta, note);
```

And leave it up to the implementer to put the properties in the proper object...for instance the method `updateTransactionGasFees` could be called with an object that look like this

```
const toUpdate = {
   userFeeLevel: "medium",
   estimatedBaseFee: "21",
   txParams: {
     "gas": "0x5208",
     "maxFeePerGas": "0x59682f2a"
   }
}
```

But to `strictly enforce that only those properties can be updated` I need to know this structure.

Reply:
@segun `gasLimit` and `gasPrice` are part of `txParams`, all of the other properties you listed are not